staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -346,15 +346,18 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 
 	if useWatchList {
 		w, err = r.watchList(stopCh)
-		if w == nil && err == nil {
-			// stopCh was closed
-			return nil
-		}
-		if err != nil {
+		switch {
+		case err != nil:
 			klog.Warningf("The watchlist request ended with an error, falling back to the standard LIST/WATCH semantics because making progress is better than deadlocking, err = %v", err)
 			fallbackToList = true
 			// ensure that we won't accidentally pass some garbage down the watch.
 			w = nil
+		case w == nil:
+			// stopCh was closed
+			return nil
+		default:
+			// Ensure watcher is stopped
+			defer w.Stop()
```

Comment:
This PR is mixing a bunch of refactoring with changes to the actual logic.
Can you please split those into separate PRs?

Reply:
Thanks, I wanted to ask for the same.

Reply:
I don't think I can really fix the bug without some of the refactoring. But I can try to pull out the refactor as separate from the bug fix.

Reply:
https://github.com/kubernetes/kubernetes/pull/125391

Reply:
Refactoring is merged and this is rebased.